Dot Exporter and Write to a file

palaniappan1 opened the pull request.

palaniappan1 committed:
Dot Exporter and Write to a file

palaniappan1 committed:
fmt format commit

JonasKlauke commented on sootup.callgraph/src/main/java/sootup/callgraph/GraphBasedCallGraph.java:
```diff
@@ -23,10 +23,11 @@
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.*;
```
Change import java.util.*; back to 
import java.util.Comparator;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

JonasKlauke commented on sootup.callgraph/src/main/java/sootup/callgraph/CallGraph.java:
```diff
@@ -86,6 +86,9 @@ boolean containsCall(
   int callCount();
 
+  /** This method converts the call graph object into dot format and write it to a string file. */
+  void convertToDotFormatAndFileWrite(String filePath);
```
exportAsDot() is a better name

swissiety replied:
is there a method which returns it as String (as well)?

palaniappan1 replied:
Now we changed the method to just return a dot formatted string and the file write feature is removed.

JonasKlauke commented on sootup.callgraph/src/main/java/sootup/callgraph/GraphBasedCallGraph.java:
```diff
@@ -129,6 +130,84 @@ public int callCount() {
+      Vertex targetVertex = graph.getEdgeTarget(edge);
+      dotFormatBuilder
+          .append("\t")
+          .append(sourceVertex.methodSignature)
```
put it in "

JonasKlauke commented on sootup.callgraph/src/main/java/sootup/callgraph/GraphBasedCallGraph.java:
```diff
@@ -129,6 +130,84 @@ public int callCount() {
+          .append("\t")
+          .append(sourceVertex.methodSignature)
+          .append(" -> ")
+          .append(targetVertex.methodSignature)
```
put it in ""

JonasKlauke commented on sootup.callgraph/src/main/java/sootup/callgraph/GraphBasedCallGraph.java:
```diff
@@ -129,6 +130,84 @@ public int callCount() {
+          .append(";\n");
+    }
+    // Sort the callGraph
+    String sortedCallGraph = sortCallGraph(dotFormatBuilder.toString());
```
dont sort the string. Sort the edges in the call graph it is way easier

JonasKlauke replied:
wasting execution time to extract strings if we have already ways to get them before we call tostring on the method signature

JonasKlauke commented on sootup.callgraph/src/main/java/sootup/callgraph/GraphBasedCallGraph.java:
```diff
@@ -129,6 +130,84 @@ public int callCount() {
+            if (className1.equals(className2)) {
+              String methodName1 = extractMethodName(line1);
+              String methodName2 = extractMethodName(line2);
+              return methodName1.compareTo(methodName2);
```
missing comparision of parameters

palaniappan1 committed:
Changes done to DotExporter

1. Removed the file write and now the method returns the DotFormat string
2. All methodsignatures are enclosed in ""
3.Unwanted imports removed
4.Sorted based on edges and parameter comparison added

JonasKlauke commented on sootup.callgraph/src/main/java/sootup/callgraph/GraphBasedCallGraph.java:
```diff
@@ -129,6 +133,93 @@ public int callCount() {
+    String sortedCallGraph = sortCallGraph(dotFormatBuilder.toString());
+
+    // Using only one string builder, so deleting everything and adding the sorted callgraph
+    dotFormatBuilder.delete(0, dotFormatBuilder.capacity());
```
Don't do this. 
Instead just return the string concatination 
"strict digraph ObjectGraph {\n" + sortedCallgraph + "}"

JonasKlauke replied:
deleting the string builder seems unnecessary complex, and it might waste execution time

JonasKlauke commented on sootup.callgraph/src/main/java/sootup/callgraph/GraphBasedCallGraph.java:
```diff
@@ -129,6 +133,93 @@ public int callCount() {
+   * @param dotOutput
+   * @return
+   */
+  public String sortCallGraph(String dotOutput) {
```
why do you still use this method?

JonasKlauke replied:
your edges are already kind of sorted. why sort them again?

JonasKlauke commented on sootup.callgraph/src/main/java/sootup/callgraph/GraphBasedCallGraph.java:
```diff
@@ -129,6 +133,93 @@ public int callCount() {
+    graph.edgeSet().stream()
+        .sorted(
+            Comparator.comparing(
+                    (Edge edge) -> graph.getEdgeSource(edge).toString(), Comparator.naturalOrder())
```
dont sort the entire signature.
otherwise the order will be class name , return type, method name and then parameter.
but we dont want that the return type highly effects the order. So you have to compare the parts of the method signatures

palaniappan1 committed:
DotExporter Changes

1. sortCallGraph() method removed
2. The comparator is written in such a way because we don't have access to methodSignature at first like in toString() method.

JonasKlauke reviewed (approved):
Nice work 👍
